dss/api/files.py

```diff
@@ -45,8 +52,100 @@ def get(uuid: str, replica: str=None, timestamp: str=None):
 
     return response
 
+
 def list():
     return dict(files=[dict(uuid=str(uuid.uuid4()), name="", versions=[])])
 
+
 def put(uuid: str):
-    return jsonify(dict(timestamp="2017-06-01T19:21:17.068Z")), requests.codes.created
+    uuid = uuid.lower()
+
+    # get the metadata to find out what the eventual file will be called.
+    request_data = request.get_json()
+
+    source_url = request_data['source_url']
+    cre = re.compile(
+        "^"
+        "(?P<schema>(?:s3|gs|wasb))"
+        "://"
+        "(?P<bucket>[^/]+)"
+        "/"
+        "(?P<object_name>.+)"
+        "$")
+    mobj = cre.match(source_url)
+    if mobj and mobj.group('schema') == "s3":
+        handle, hca_handle, dst_bucket = \
+            Config.get_cloud_specific_handles("AWS")
+    else:
+        # TODO: (ttung) better error messages pls.
+        return (
+            make_response("I can't support this source_data schema!"),
+            requests.codes.bad_request,
+        )
+
+    src_bucket = mobj.group('bucket')
+    src_object_name = mobj.group('object_name')
+
+    metadata = handle.get_metadata(src_bucket, src_object_name)
+
+    # what's the target object name for the actual data?
+    dst_object_name = ("blobs/" + ".".join(
+        (
+            metadata['hca-dss-sha256'],
+            metadata['hca-dss-sha1'],
+            metadata['hca-dss-s3_etag'],
+            metadata['hca-dss-crc32c'],
+        )
+    )).lower()
+
+    # does it exist? if so, we can skip the copy part.
+    try:
+        handle.get_metadata(
+            dst_bucket, dst_object_name)
+    except BlobNotFoundError:
+        hca_handle.copy_blob_from_staging(
+            src_bucket, src_object_name,
+            dst_bucket, dst_object_name)
```

Comment:
Consider rolling these into one operation, by using a put with a precondition (I think the only one you could use for this is `if-unmodified-since`, but I'm not 100% sure).